### ISSUE TITLE
Unclear incorporation of user or item metadata

### ISSUE THREAD

**Issue by EdwardALockhart** [tensorflow/recommenders#565]
Hi,

I have been going through the tutorials and one question remains which is the incorporation of user and item metadata.

I have seen examples where both user and item metadata are incorporated into the User Model (Query Model - the interchangeable names can be confusing). For example here https://github.com/drtinumohan/tfrs_amazon_dataset/blob/main/tfrs_amazon.ipynb, there are user attributes (residence) and item attributes (cabin type) being incorporated into the User Model. Why not incorporate them both or one of them in the Item Model (Candidate Model)?

Is there any standard approach detailing where such features should be incorporated and how they might fit into say the basic quick start example here? https://www.tensorflow.org/recommenders/examples/quickstart

I know that both layers will have to have vocabs produced to convert to integers, but beyond that, what rules should be followed in terms of slotting these metadata into a simple retrieval model like in the quick start example?

Thanks!

**Comment by patrickorlando** [comment]
I would suggest you think in terms of queries and candidates.
A query is the context for which you want to retrieve a set of candidates. The query input features **should not** contain information about the target candidate that is not available at inference time.

In the example above, the user is performing a search. The query is the `USER_ID`, the user's country `USER_RESIDENCE` and the desired travel class `CABIN_TYPE`. All of this can be provided at inference time.

So the distinction is pretty clear. You can add information about the user or the recommendation context to the query tower and you can add item features to the candidate tower.

**Comment by EdwardALockhart** [comment]
I see, this makes a lot more sense now when you try to write the code and get recommendations for a specific user - you also have to supply the other characteristics if you added to to the query model, some of which you couldn't possibly know at inference time.

Thanks!

**Comment by EdwardALockhart** [comment]
So I can bring across the multi-item field into the table by simply concatenating as below. Does this seem sensible?
```
index = tfrs.layers.factorized_top_k.BruteForce(retrieval_model.query_model)
index.index_from_dataset(items.batch(100).map(lambda x: (x['item'] + x['item_type'],
                                                         retrieval_model.candidate_model(x))))
```

I can get out a recommendation as item + item_type for a single user using
```
user = "user"
affinity_scores, recommended_items = index(tf.constant([user]))
recommended_items = recommended_items.numpy().tolist()[0]
```
Just working out how to do it for user + user_type instead of just user in the example above

**Comment by patrickorlando** [comment]
Yep, exactly. Thinking about how you intend to use the model typically can clear up what features you can use and where.

You are misunderstanding the index_from_dataset method.

https://github.com/tensorflow/recommenders/blob/8b249f3fc0f8d3d907eecf010809a5df3759d65d/tensorflow_recommenders/layers/factorized_top_k.py#L184-L189

You need to map your dataset so that it consists of tuples `(item_id, item_vector)`, as follows.
```python
index = tfrs.layers.factorized_top_k.BruteForce(retrieval_model.query_model)
index.index_from_dataset(items.batch(100).map(
    lambda inputs: (inputs['item'], retrieval_model.candidate_model(inputs))
))
```

**Comment by EdwardALockhart** [comment]
Thanks for this. Because I'm supplying metadata with my items in the form or item + item_type... when I map items like you suggest then the index obviously contains duplicates. So for a given query the same item can appear multiple times as the item_type is omitted which would have allowed their differentiation.

Is there no way to supply items with metadata into the model that you can get just unique items out? Or do you have to supply metadata (fudging the index_from_dataset bit as I did) or just supply items as you have suggested and de-duplicate the list?

**Comment by patrickorlando** [comment]
The id is used for nothing more than identifying the recommended item. It should be unique for each candidate. I don't fully understand how you can have the same item with different type, but if you need to create a new composite id to make it work then that's fine.
